File: scenarios/typescript-code-graph/sidecar/src/handshake.ts
// Version negotiation. The Go supervisor sends {type:"handshake",
// protocol_version, request_id}; the sidecar replies with its own
// protocol_version, sidecar version, and ts-morph version. The supervisor
// rejects mismatched major versions (handled Go-side).

import { createRequire } from "node:module";

import type { HandshakeRequest, HandshakeResponse } from "./protocol.js";
import { PROTOCOL_VERSION } from "./protocol.js";

const require = createRequire(import.meta.url);

let cachedTsMorphVersion: string | null = null;
function tsMorphVersion(): string {
  if (cachedTsMorphVersion !== null) return cachedTsMorphVersion;
  try {
    const pkg = require("ts-morph/package.json") as { version: string };
    cachedTsMorphVersion = pkg.version;
  } catch {
    cachedTsMorphVersion = "unknown";
  }
  return cachedTsMorphVersion;
}

let cachedSelfVersion: string | null = null;
function selfVersion(): string {
  if (cachedSelfVersion !== null) return cachedSelfVersion;
  try {
    const pkg = require("../package.json") as { version: string };
    cachedSelfVersion = pkg.version;
  } catch {
    cachedSelfVersion = "0.0.0";
  }
  return cachedSelfVersion;
}

export function handleHandshake(req: HandshakeRequest): HandshakeResponse {
  return {
    type: "handshake",
    request_id: req.request_id,
    protocol_version: PROTOCOL_VERSION,
    sidecar_version: selfVersion(),
    ts_morph_version: tsMorphVersion(),
  };
}

File: scenarios/typescript-code-graph/api/internal/sidecar/handshake.go
package sidecar

// ProtocolVersion is the IPC protocol major version this Go side
// speaks. The sidecar must echo the same number in its handshake_ok
// reply; any other value is a fatal mismatch.
const ProtocolVersion = 1

// handshakeRequest / handshakeResponse mirror plan §8.4 wire shapes.
type handshakeRequest struct {
	Type            string `json:"type"`
	RequestID       string `json:"request_id"`
	ProtocolVersion int    `json:"protocol_version"`
}

type handshakeResponse struct {
	Type            string `json:"type"`
	RequestID       string `json:"request_id"`
	ProtocolVersion int    `json:"protocol_version"`
	SidecarVersion  string `json:"sidecar_version"`
}
